solr/contrib/blob-directory/src/java/org/apache/solr/blob/BlobDirectoryFactory.java

```diff
@@ -20,31 +20,49 @@
 import java.io.File;
 import java.io.IOException;
 import java.lang.invoke.MethodHandles;
+import java.net.URI;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Locale;
 import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 import org.apache.commons.io.FileUtils;
 import org.apache.lucene.store.*;
 import org.apache.solr.common.SolrException;
+import org.apache.solr.common.params.CoreAdminParams;
 import org.apache.solr.common.params.SolrParams;
 import org.apache.solr.common.util.IOUtils;
 import org.apache.solr.common.util.NamedList;
 import org.apache.solr.core.CachingDirectoryFactory;
 import org.apache.solr.core.CoreContainer;
 import org.apache.solr.core.CoreDescriptor;
 import org.apache.solr.core.DirectoryFactory;
+import org.apache.solr.core.backup.repository.BackupRepository;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class BlobDirectoryFactory extends CachingDirectoryFactory {
 
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
+  /**
+   * Default maximum number of threads that push files concurrently to the backup repository.
+   */
+  private static final int DEFAULT_MAX_THREADS = 20;
+
+  /**
+   * Default size of the buffer used to transfer input-output stream to the {@link BackupRepository}, in bytes.
+   * The appropriate size depends on the {@link BackupRepository}.
+   * There is one buffer per thread.
+   */
+  public static final int DEFAULT_STREAM_BUFFER_SIZE = 32_768;
+
   private static final Pattern INDEX_NAME_PATTERN = Pattern.compile("index(?:\\.[0-9]{17})?");
 
   private String localRootPath;
-  private BlobStore blobStore;
+  private BackupRepository backupRepository;
```

Comment:
In terms of naming, let's just use the word "repository" to refer to the remote place?  I know of course that the abstraction is "BackupRepository" but the "backup"-ness aspect could plausibly be renamed away so that it's clearer that it's more general than backups.  I mentioned this in JIRA somewhere.

Reply:
I see "BlobStore" in some places that could be simply "Repo" or "Repository"

Reply:
+1